### ISSUE TITLE
Spring WebFlux reactive threads hanging, required Jetty 9.4.x

### ISSUE THREAD

**Issue by ati90ati** [tomakehurst/wiremock#914]
Wiremock is using a version of Jetty 9.2.x and for the new version of Spring Boot with the WebFlux (2.0.0.RELEASE) Reactive threads sometimes the calling thread stops and hanging when calling Wiremock. This problem occurs only in version 9.2.x and not in 9.4.8.v20171121 of Jetty.

I understand that you want the JDK 7 support, but in that case you'll lose support for newer technologies/frameworks like Spring WebFlux, HTTP2 and so on...
Wiremock with version 9.2.x of Jetty is unusable with reactive threads because it randomly stops hanging.

I saw your conversation on https://github.com/tomakehurst/wiremock/pull/887 thread and the commit https://github.com/tomakehurst/wiremock/commit/4368d316ec239bac305f1faadedab24765bbe4b6, but somehow it will be very nice if you can provide support for the new Spring 5 WebFlux (Spring Boot WebFlux 2.0.0.RELEASE).

> Add another JAR artifact to the build that's Java 8+ only and uses the latest Jetty version. This would probably involve splitting the source tree up into core and version-specific parts, which the build would include/exclude for the various output JARs.

> Add some runtime detection of the environment i.e. are we on Java 8+ and Jetty 9.4+, and load the HTTP/2 code only under these conditions, such that NoClassDefFoundErrors and the like aren't thrown when running on Java 7.

Any of the above solution should be fine. The idea is to upgrade also the Wiremock Standalone which is used by spring-cloud-contract-wiremock.

Please see related issues:
https://github.com/spring-projects/spring-boot/issues/10569
https://github.com/reactor/reactor-netty/issues/264
https://github.com/spring-cloud/spring-cloud-contract/issues/606

**Comment by ati90ati** [comment]
Isolating it to Jetty for me seems that I've tried with 9.2.22.v20170606 was hanging and with 9.4.8.v20171121 not.

I know that the bug what I've added there is from Reactor Netty, but I think that some similar bug existed also in Jetty, because with 9.2.x randomly hangs and for 9.4.x don't.

Please check this comment:
https://github.com/spring-cloud/spring-cloud-contract/issues/606#issuecomment-379725684

I've tried, and found out that for me only work with normal WireMock and Jetty 9.4.x.
With WireMock Standalone and Jetty 9.4.x doesn't worked.

Also https is not working with normal WireMock and Jetty 9.4.x.

**Comment by tomakehurst** [comment]
I can see from the linked ticket that you're using the spring-cloud-wiremock dependency, but are you actually initialising WireMock in your code using the annotation? Or are you using WireMock's rule or server classes?

**Comment by ati90ati** [comment]
I'm using the `@AutoConfigureWireMock(port = 0)` annotation.
And in application.yml setting the url like this:

    api:
        my-other-service.url: http://localhost:${wiremock.server.port}

**Comment by tomakehurst** [comment]
I'm quite puzzled as to why you're seeing this problem at all in that case. You shouldn't have been using WireMock's version of Jetty either way.

**Comment by OlgaMaciaszek** [comment]
Since we have started using WireMock-standalone as our dependency, we take the Jetty version from WireMock.

**Comment by jmayday** [comment]
> 
> Wiremock with version 9.2.x of Jetty is unusable with reactive threads because it randomly stops hanging.

Well, I just noticed my tests randomly hang while using wiremock-jre8 2.21.0 in Spring Boot 2.1.0.RELEASE. Output looks like this:

```
2019-03-06 10:26:05.036  INFO 13136 --- [qtp399685190-67] o.e.j.s.handler.ContextHandler.ROOT      : RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.StubRequestHandler. Normalized mapped under returned 'null'
2019-03-06 10:26:05.043  INFO 13136 --- [qtp399685190-67] WireMock                                 : Request received:
127.0.0.1 - GET /url/goes/here

User-Agent: [ReactorNetty/0.8.2.RELEASE]
Host: [localhost:58294]
Accept-Encoding: [gzip]
Accept: [application/json;charset=UTF-8]
Content-Type: [application/json]



Matched response definition:
{
  "status" : 200,
  "body" : "{\"data\":\"value"}",
  "headers" : {
    "Content-Type" : "application/json"
  }
}

Response:
HTTP/1.1 200
Content-Type: [application/json]
Matched-Stub-Id: [47ef99f4-1788-4a78-9cb1-8610e925b6b1]


2019-03-06 10:26:05.044  INFO 13136 --- [ctor-http-nio-3] <class>      : 200 OK GET http://localhost:58294/url/goes/here took 27ms
```

**Comment by jmayday** [comment]
It seems that in my case tests randomly hang both on Windows and Linux machines. But the difference is that on Linux machine they correctly timeout (and execution of tests is continued) while on Windows they hang forever. This ticket has been referenced: https://github.com/eclipse/jetty.project/issues/2205

**Comment by tomakehurst** [comment]
@jmayday have you tried upgrading to `wiremock-jre8`, which has the latest Jetty version?

**Comment by jmayday** [comment]
> @jmayday have you tried upgrading to `wiremock-jre8`, which has the latest Jetty version?

Yes, as I stated in previous post - I'm already using wiremock-jre8 dependency (in newest version available - 2.21.0)

**Comment by tomakehurst** [comment]
Sorry, I saw that just after I'd submitted the post, which is why I deleted it!

**Comment by GideonLiem** [comment]
We've encountered the same problem.
When we use wiremock to test our webflux code, about 1 out of 4 times, it will cause the test to hang and wait indefinitely. 

Does anybody have a fix or workaroud for this?

**Comment by tomakehurst** [comment]
@GideonLiem I'm working on upgrading to the latest Jetty version, hopefully for release in the next few days.

However, I don't know whether this will make any difference. I did a little bit of analysis on this (see https://github.com/tomakehurst/wiremock/issues/1084) and I suspect there may be a race condition in the `WebClient` handling of pooled connections.

You might find you can work around the problem by turning off connection pooling in your client when testing, or setting a `Connection: close` header.

**Comment by joakime** [comment]
@tomakehurst (jetty dev here) out of curiosity, is any of the testing that is hanging dealing with a Client using `Transfer-Encoding: chunked`?

**Comment by tomakehurst** [comment]
@joakime Yes, the test linked from #1084 involves `Transfer-Encoding: chunked` responses

**Comment by joakime** [comment]
@tomakehurst take a close look at the reading of the input streams when using `Transfer-Encoding: chunked`, as libraries (such as json) often do not read to the -1/EOF.
That's a problem when you combine HTTP/1.1 and `Transfer-Encoding: chunked` as not reading to EOF for the inputstream means the Connection is in an invalid state and cannot be reused for subsequent requests (persistent connections and HTTP/1.1).

The fact that people have pointed out that String works, but JSON does not, seems to indicate that the JSON implementation isn't reading to EOF on a Chunked transfer.

**Comment by joakime** [comment]
The fact that this is also reported as random fits really well into this `Transfer-Encoding: chunked` behavior.
Due to networking buffers and whatnot, the final "chunk" could be part of the fill/read sometimes, and sometimes it's in another fill/read pending another read from the application.

**Comment by clement-buchart** [comment]
Hello, 

We had the same issue, and managed to solves it (at least we didn't reproduce since the workaround) via this Transformer, that explicitly adds "Connection: Close" header :

```
public class ConnectionCloseExtension extends ResponseTransformer {

    @Override
    public Response transform(Request request, Response response, FileSource files, Parameters parameters) {
        return Response.Builder
                .like(response)
                .headers(HttpHeaders.copyOf(response.getHeaders())
                        .plus(new HttpHeader("Connection", "Close")))
                .build();
    }

    @Override
    public String getName() {
        return "ConnectionCloseExtension";
    }
}

```

(inspired by discussion at https://github.com/tomakehurst/wiremock/issues/485)

Hope this helps going further with the diagnostic.

**Comment by GideonLiem** [comment]
> @GideonLiem I'm working on upgrading to the latest Jetty version, hopefully for release in the next few days.
> 
> However, I don't know whether this will make any difference. I did a little bit of analysis on this (see #1084) and I suspect there may be a race condition in the `WebClient` handling of pooled connections.
> 
> You might find you can work around the problem by turning off connection pooling in your client when testing, or setting a `Connection: close` header.

Thanks for your reply.
We have tried the Connection: close header already but it didn't make any difference.

**Comment by tomakehurst** [comment]
I've learnt two interesting things about this problem today.

1) Creating a very large, fixed connection pool for the `WebClient` fixes the test in #1084 e.g.
```java
WebClient client = WebClient.builder()
                    .clientConnector(new ReactorClientHttpConnector(
                            HttpClient.create(ConnectionProvider.fixed("blah", 5000))
                    ))
                    .build();
```

2) Sending a `Content-Length` header in WireMock responses (and hence not using chunked transfer encoding - good shout @joakime) also fixes the tests.

I'm going to add a configuration switch or stub element (or both) in the next version of WireMock to specify whether you want responses to be chunk encoded.

**Comment by tomakehurst** [comment]
I still don't know what this means in terms of root cause. My best guess is that `WebClient` struggles with chunked responses when it has to close the connection, but there's a chance Jetty is (also) doing something dodgy here.

**Comment by GideonLiem** [comment]
> I still don't know what this means in terms of root cause. My best guess is that `WebClient` struggles with chunked responses when it has to close the connection, but there's a chance Jetty is (also) doing something dodgy here.

We are actually doing our tests with TestRestTemplate.. so this problem not only occurs for WebClient.

**Comment by tomakehurst** [comment]
@GideonLiem which underlying HTTP client are you using with `TestRestTemplate`?

Do you have a failing test project you could share?

**Comment by GideonLiem** [comment]
> @GideonLiem which underlying HTTP client are you using with `TestRestTemplate`?
> 
> Do you have a failing test project you could share?

I am using the standard one for Spring TestRestTemplate. No change in configuration.
I don't have a failing test project available, since the code is part of a big project which is confidential.

**Comment by tomakehurst** [comment]
OK, so it's probably defaulting to Java's `HttpURLConnection` as its backend. It might be worth you switching it to use the Apache HTTP client implementation instead. This works in some cases.

**Comment by ati90ati** [comment]
Hello guys,
I succeed to create an example project where using the `WebClient` from `Webflux` the tests with `Wiremock` randomly fails.

https://github.com/ati90ati/bugreport-wiremock-webflux-hanging

I'm getting this error:

```
java.lang.IllegalStateException: Timeout on blocking read for 5000 MILLISECONDS

	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:117)
	at reactor.core.publisher.Mono.block(Mono.java:1518)
	at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultRequestBodyUriSpec.exchange(DefaultWebTestClient.java:283)
	at org.anagy.bugreport.controller.NumberControllerIntegrationTest.testGetNumbers(NumberControllerIntegrationTest.java:62)
```

Initially I wanted to have 10 services but I observed that running the tests many times even with one service randomly fails, so at now I've decided to leave 2 services.

Sometimes it fails after 218 tests, sometimes does not fail, sometimes only after a few thousands.

If it's needed I can add / refine the tests to have it more systematic (more tests less upper range) or add more services to have more mocks.

NOTE: The tests in our bigger project failed even using Java 8, even with Java 11.

**Comment by tomakehurst** [comment]
I've just released WireMock 2.22.0, which includes the ability to switch between chunked encoding and `Content-Length`. I'd be interesting to see how you get on with this.

**Comment by ati90ati** [comment]
Sorry, but I cannot find the version 2.22.0 on the Maven Central repo. Even the wiremock-jre8 what I'm using in the project.

**Comment by tomakehurst** [comment]
Sometimes it takes a while to sync over from Sonatype. However, I can see it there:
https://repo1.maven.org/maven2/com/github/tomakehurst/wiremock-jre8/2.22.0/

BTW, you can grab releases straight from Sonatype if they're not yet in M2 e.g.
https://oss.sonatype.org/content/groups/public/com/github/tomakehurst/wiremock-jre8/2.22.0/

**Comment by joakime** [comment]
The sample project https://github.com/ati90ati/bugreport-wiremock-webflux-hanging can be run from command line without failure reported. (been on a loop for the past 15 hours).
I need to be able to run it from an IDE, but that project is rejected as invalid by Eclipse IDE 2018-12, and IntelliJ 2018.3.5 (CE and Ultimate) as uncompilable, so not runnable.
This would allow me to setup some debugging, and hopefully at least point to the specific behavior that is tripping you up.

**Comment by ati90ati** [comment]
I've removed the lombok plugin from the sample project so everyone use it.

Actually now I've run it 3 times and each time it failed for me.

As I observed this random failure occurs more often if the PC (CPU Frequency) used has less performance.

I've get less time failures at a Desktop PC with 3.4Ghz 8 cores, and more tie failures with a laptop 2.3Ghz (turbo boost 3.1) 8 core CPU.

For me it's failing even from IDE even from command line.

NOTE: the tests are not run in parallel.

![Screenshot from 2019-03-18 19-57-15](https://user-images.githubusercontent.com/14107633/54552822-9fc86a80-49b9-11e9-9e4f-c5b4c211c0b8.png)
![Screenshot from 2019-03-18 19-59-11](https://user-images.githubusercontent.com/14107633/54552823-9fc86a80-49b9-11e9-9ff0-071d3bc76d7c.png)
![Screenshot from 2019-03-18 20-00-45](https://user-images.githubusercontent.com/14107633/54552824-9fc86a80-49b9-11e9-9109-a853b21384c7.png)

**Comment by ati90ati** [comment]
I'll try a little bit later with the wiremock 2.22.0 version and `Content-Length`, but on the Maven Central I still cannot see the new version (I'll use the suggested from @tomakehurst ):

![image](https://user-images.githubusercontent.com/14107633/54553400-cd61e380-49ba-11e9-9e26-ecba8f65ed6a.png)

**Comment by joakime** [comment]
@ati90ati mvnrepository.com is not a great site, its often wrong (not up to date, missing entries, pointing to entries at repositories not part of the maven repository system, and generally has bad data all around)

Use the official search for maven central.
https://search.maven.org/search?q=a:wiremock

https://search.maven.org/artifact/com.github.tomakehurst/wiremock/2.22.0/pom

**Comment by ati90ati** [comment]
Hello guys, you are really awesome!

I've increased `Wiremock` version to `2.22.0` and used the `Options.ChunkedEncodingPolicy.NEVER` or `Options.ChunkedEncodingPolicy.BODY_FILE` new options and no more hanging in tests.

I've tested with 300.000 request to Wiremock and no in is remains hanging.
Before from 5000 calls for me approx 2 remains in hanging state.
I've rechecked it with the `Options.ChunkedEncodingPolicy.ALWAYS` option and it again started to hang.

I've crated a branch to show my solution, maybe I'll merge it in the future,

I've also checked the chunked encoding policy (NEVER, BODY_FILE) setting in our bigger project and seems there also to be ok, no tests hanging util now.
We can finally re-enable our integration test :smile: !

I've created a `WiremockConfig` class for just testing purposes but I'm using the `spring-cloud-contract-wiremock`.
Maybe if @tomakehurst have a little bit of time can you help to find the best solution how to put `Options` using `spring-cloud-contract-wiremock`?

Thank you really much for your help! It is really appreciated!

**Comment by GideonLiem** [comment]
> Hello guys, you are really awesome!
> 
> I've increased `Wiremock` version to `2.22.0` and used the `Options.ChunkedEncodingPolicy.NEVER` or `Options.ChunkedEncodingPolicy.BODY_FILE` new options and no more hanging in tests.
> 
> I've tested with 300.000 request to Wiremock and no one is remains hanging.
> Before from 5000 calls for me approx 2 remains in hanging state.
> I've rechecked it with the `Options.ChunkedEncodingPolicy.ALWAYS` option and it again started to hang.
> 
> I've crated a branch to show my solution, maybe I'll merge it in the future,
> 
> I've also checked the chunked encoding policy (NEVER, BODY_FILE) setting in our bigger project and seems there also to be ok, no tests hanging util now.
> We can finally re-enable our integration test 😄 !
> 
> I've created a `WiremockConfig` class for just testing purposes but I'm using the `spring-cloud-contract-wiremock`.
> Maybe if @tomakehurst have a little bit of time can you help to find the best solution how to put `Options` using `spring-cloud-contract-wiremock`?
> 
> Thank you really much for your help! It is really appreciated!

Do you have an example project to see this working with the options you mentioned?

**Comment by tomakehurst** [comment]
@ati90ati marvellous, glad we're making some progress finally! Thanks for sharing your test results.

For help getting this into Spring Cloud Contract I think we need to hail @marcingrzejszczak

**Comment by ati90ati** [comment]
@GideonLiem as I posted in my comments above there are a sample project under:
https://github.com/ati90ati/bugreport-wiremock-webflux-hanging
There is a branch named `fix-hanging` there you can find my dirty quick solution.
As I mentioned maybe I'll merge that branch in the master, so future visitors will be able to find the fix there.

**Comment by marcingrzejszczak** [comment]
Sure, can you file an issue in Spring Cloud Contract directly?

**Comment by joakime** [comment]
The bug is in the expectations on the server side.
Since the JSON parser is setup to read from the Request.inputStream/reader, it doesn't handle situations like Chunked transfer-encoding properly.
The server side should read the entire stream, what if the content submitted was `{}{}` ? would it read the second json? should it throw an error? (the most common response for this would be to respond with a status code 400 Bad Request)
From Jetty's point of view, it cannot know when the application is finished with the stream.

What happens in the failing cases with chunked transfer encoding is that the chunks arrive via the network buffers and the application (the json parser in this specific case) reads only to the final closing `}` and stops.

If the network buffers happen to contain the final closing chunk `0\r\n\r\n` then it works.
But if the network buffers that have been read to satisfy the request.inputStream has only read active chunks (but not yet the closing chunk) then the stream is still open, waiting for the application to issue another read of the connection.  (this is especially aggravated if using Async I/O features of Servlet 3.1)
If you don't finish reading, this is an "unconsumed input" scenario, and as such the HTTP/1.1 persistent connection is now in an unsatisfied state on the server side (you have a request that hasn't been finished being read, so that connection cannot be reused for another request).
The combination of the server side not reading till EOF (aka the final closing chunk) and the HTTP Client not realizing this state making the followup request that the HTTP Client is sending on that same connection not able to be processed by Jetty. Hence the timeout you see.

Proper REST endpoints will read the entire stream and if there is anything amiss about the content (eg: there is non-whitespace after the first JSON body content) it will fail the request.

**Comment by tomakehurst** [comment]
Thanks for your analysis @joakime.

One thing I'm not totally clear about - you're saying the problem is on the server side due to the JSON parser, but in the test linked to from #1084 the server is only handling GETs and isn't doing any JSON parsing. Since switching away from chunked encoding for responses fixes the problem, it seems somewhat unlikely to me that server-side request handling is the problem.

Perhaps you were referring to the JSON parsing on the client?

**Comment by GideonLiem** [comment]
The **solution** that brought the fix for us was writing our own extension. After that we hhad no more problems.

```
public class CustomResponseTransformExtension extends ResponseTransformer {
    @Override
    public Response transform(Request request, Response response, FileSource files, Parameters parameters) {
        return Response.Builder
                .like(response)
                .headers(HttpHeaders.copyOf(response.getHeaders())
                        .plus(new HttpHeader("Connection", "close"),
                                new HttpHeader("Content-Encoding", "none")))
                .build();
    }

    @Override
    public String getName() {
        return "ConnectionCloseExtension";
    }
}
```


And using the extension in our Spock integration test:

```
@Shared
@ClassRule
WireMockClassRule wireMockRule = new WireMockClassRule(WireMockConfiguration.wireMockConfig()
        .extensions(CustomResponseTransformExtension).port(WIREMOCK_PORT))
```

**Comment by tomakehurst** [comment]
@GideonLiem so it turned out that disabling chunked encoding wasn't sufficient?

**Comment by GideonLiem** [comment]
@tomakehurst Yes. We tried that as well.. and we still got the same problems.

**Comment by tomakehurst** [comment]
@GideonLiem could you share a test project that illustrates this? In the project linked to from this thread, disabling chunked response encoding fixed it, so I'm not sure how to go about replicating now.

**Comment by GideonLiem** [comment]
Unfortunately I don't have a sample test project with isolated problem available. But I thought it was fair to at least share our solution ;-)

**Comment by tomakehurst** [comment]
We've seen disabling of connection reuse work in the past, but I haven't been able to isolate the circumstances in which it's the correct solution.

Also, it shouldn't ever be necessary (you'd never consider disabling it in a production system). It either indicates a server or client bug IMO, but without a reliable failing test I can't investigate which it is.

**Comment by joakime** [comment]
Even a wireshark/tcpdump of the failing scenario would be useful.

**Comment by be-hase** [comment]
I think this issue is fixed by https://github.com/netty/netty/pull/8922

**Comment by timpeeters** [comment]
> I think this issue is fixed by [netty/netty#8922](https://github.com/netty/netty/pull/8922)

I'm still getting the same error even with the netty/netty#8922 fix included in Netty 4.1.34.Final on the classpath.

**Comment by be-hase** [comment]
Really?
With `netty 4.1.34`, this problem no longer reproduces.
@ati90ati 's example project (https://github.com/ati90ati/bugreport-wiremock-webflux-hanging)

**Comment by kimec** [comment]
@tomakehurst WebFlux/reactor-netty connection pooling is brittle, especially in combination with Jetty 9.4 server. We had this setup in production and it did not work. In particular, reactor-netty does not handle racy close(from server)/write (from client) scenario well - it actually does not handle it at all (reactor/reactor-netty#612). If you have just the right workload, you can run in racy close(from server)/write (from client) scenario very often.
I think Jetty 9.4 likes to kill idle connections after 30 seconds by default while it seems like reactor-netty would love to have idle connections around forever.

@timpeeters Have you tried Jetty client based WebFlux client from SB 2.1/Spring 5.1?

**Comment by timpeeters** [comment]
@kimec Yes, we are running SB 2.1.4.RELEASE which depends on Spring 5.1.6.RELEASE. Dependency tree looks clean (no version conflicts) and the latest WireMock version 2.23.2.

However, I do need to add that we run Spring Cloud Gateway on top of SB. We mainly run into this issue during integration tests using WireMock as a stub for our backend services.

**Comment by kimec** [comment]
@timpeeters I see. Have you also configured your WebFlux client connecting to WireMock [to use JettyClientHttpConnector](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#webflux-client-builder-jetty)?

**Comment by timpeeters** [comment]
@kimec no, we do not use the JettyClientHttpConnector. We use the default WebFlux client created by Spring Cloud Gateway in GatewayAutoConfiguration > NettyConfiguration > httpClient().
It does not use the Jetty connector.

**Comment by oberlies** [comment]
> I think this issue is fixed by netty/netty#8922

Yessss!!! A was getting at least one case of WebClient&WireMock hanging whenever I ran our integration test suite, and with a Spring Boot update from 2.1.3.RELEASE to 2.1.4.RELEASE (which entails netty 4.1.33.Final -> 4.1.34.Final), the problem is gone! :-)

**Comment by tomakehurst** [comment]
@oberlies that's great news, thanks for sharing!

**Comment by ati90ati** [comment]
I can confirm that this bug has two solutions now.

**Solution 1:**
With Spring Boot 2.1.4.RELEASE the problem disappears.


**Solution 2:**
With using at least Wiremock 2.22.0 version and the wiremock options `ChunkedEncodingPolicy.BODY_FILE` the problem also can be fixed.
This solution with chunked body is described here: https://github.com/tomakehurst/wiremock/issues/914#issuecomment-474262581
Example config with static port:
```java
@Configuration
public class WiremockConfig {
    @Bean
    public Options wiremockOptions() {
        final com.github.tomakehurst.wiremock.core.WireMockConfiguration wireMockConfiguration = WireMockSpring
                .options();

        wireMockConfiguration.useChunkedTransferEncoding(Options.ChunkedEncodingPolicy.BODY_FILE);
        wireMockConfiguration.port(12345);

        return wireMockConfiguration;
    }
}
```

So, the issue can be closed.

**Comment by tomakehurst** [comment]
@ati90ati thanks for sharing

**Comment by chetz3** [comment]
does the Flux stream close automatically and frees the thread for the next call in this case 
`}).flatMapMany(sess -> {
            InputStreamReader inputStreamReader = new InputStreamReader(sess.getStdout());
            BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
            connectionMap.put(inputStream, inputStreamReader);
            connectionMap.put(bufferedR, bufferedReader);
            Flux<Long> interval = Flux.interval(Duration.ofMillis(50));
            return stream ? interval.map(i -> {
                try {
                    return bufferedReader.ready() ? bufferedReader.readLine() : "";
                } catch (Exception ignore) {
                }
                return "";
            }) : Flux.zip(Flux.fromStream(bufferedReader.lines()), interval, (key, value) -> key);
        }).doOnTerminate(runnable).doOnCancel(runnable).doOnComplete(runnable).doOnError(throwable -> {`
i see the application is hanging on opening multiple sessions

**Comment by mkopylec** [comment]
I'm using `compile group: 'com.github.tomakehurst', name: 'wiremock-jre8', version: '2.26.3'` which contains dependency to jetty 9.4.20. I have this problem too: my tests randomly hang. When I turn off wiremock (not start it) in my tests everything runs without any problem (thus tests don't pass but that is obviuos)

**Comment by tomakehurst** [comment]
@mkopylec are you using a version of Netty newer than `4.1.34` - the one that fixes the bug causing this?

If so, please can you provide a failing test project.

**Comment by mkopylec** [comment]
I'm using Netty `4.1.43`, I can't provide a test project cause it is an internal company project :(

**Comment by tomakehurst** [comment]
Not a lot I can do about it then I'm afraid 🤷‍♂️
